scripts/calculateNetworkRevenue.ts

```diff
@@ -0,0 +1,168 @@
+import { Address, stringToHex } from 'viem'
+import yargs from 'yargs'
+import { NetworkId, protocols } from './types'
+import { getRegistryContract } from './utils'
+import {
+  NETWORK_ID_TO_HYPERSYNC_URL,
+  NETWORK_ID_TO_REGISTRY_ADDRESS,
+} from './utils/networks'
+import {
+  HypersyncClient,
+  QueryResponse,
+  TransactionField,
+} from '@envio-dev/hypersync-client'
+
+async function main(args: ReturnType<typeof parseArgs>) {
+  const networkId = args.networkId as NetworkId
+  const protocolId = args['protocol-id'] as string
+  const startBlock = (args['start-block'] as number) ?? 0
+  const endBlock = (args['end-block'] as number) ?? null
+
+  if (!NETWORK_ID_TO_HYPERSYNC_URL[networkId]) {
+    console.log(`Network ID ${networkId} is not supported by HyperSync`)
+    return
+  }
+
+  const client = HypersyncClient.new({
+    url: NETWORK_ID_TO_HYPERSYNC_URL[networkId],
+    bearerToken: process.env.HYPERSYNC_API_KEY,
+  })
+
+  const users = await getUsersForProtocol({ networkId, protocolId })
+  if (users.length === 0) {
+    console.log(`No users found for protocol ${protocolId} on ${networkId}`)
+    return
+  }
+
+  const totalGasUsedWei = await fetchTotalGasUsed({
+    client,
+    users,
+    startBlock,
+    endBlock,
+  })
+  console.log(`Total gas used (Wei): ${totalGasUsedWei}`)
+}
+
+async function fetchTotalGasUsed({
+  client,
+  users,
+  startBlock,
+  endBlock,
+}: {
+  client: HypersyncClient
+  users: { userAddress: string }[]
+  startBlock: number
+  endBlock: number | null
+}): Promise<bigint> {
+  const userAddresses = users.map((user) => user.userAddress)
+
+  const query = {
+    transactions: [{ from: userAddresses }],
+    fieldSelection: {
+      transaction: [TransactionField.GasUsed, TransactionField.GasPrice],
+    },
+    fromBlock: startBlock,
+    ...(endBlock !== null && { toBlock: endBlock }),
+  }
+
+  try {
+    const response: QueryResponse = await client.get(query)
+    return response.data.transactions.reduce(
+      (sum: bigint, tx) =>
+        sum + BigInt(tx.gasUsed ?? 0) * BigInt(tx.gasPrice ?? 0),
+      0n,
+    )
+  } catch (error) {
+    console.log('Error fetching gas usage: ', error)
+    return 0n
+  }
+}
+
+async function getUsersForProtocol({
```

Comment:
can you extend the existing fetchReferrals script instead of writing a new one?

Reply:
I was able to use the functions from `fetchReferralEvents` and `removeDuplicates` instead of creating my own to avoid duplication.